File: Algorithm_Mentoring/25-2/python/4주차/별_찍기_10.py
import sys
sys.setrecursionlimit(10**7)
input = sys.stdin.readline

MAX = 7010
a = [[False] * MAX for _ in range(MAX)]

def dnc(cur, cy, cx):
    # Base Case
    if cur == 1:
        a[cy][cx] = True
        return

    nxt = cur // 3
    for i in range(3):
        for j in range(3):
            # 가운데면 건너 뜀
            if i == 1 and j == 1:
                continue

            # 분할
            dnc(nxt, cy + i * nxt, cx + j * nxt)

n = int(input())
dnc(n, 0, 0)

for i in range(n):
    for j in range(n):
        print('*' if a[i][j] else ' ', end = '')
    print()

File: Algorithm_Mentoring/25-2/cpp/4주차/별_찍기_10.cpp
#include <iostream>
using namespace std;
using ll = long long;

const ll MAX = 7010;
ll n, m, k;
bool a[MAX][MAX];

void dnc(ll cur, ll cy, ll cx){
    // Base Case
    if(cur == 1){
        a[cy][cx] = 1;
        return;
    }

    for(int i = 0;i < 3;i++){
        for(int j = 0;j < 3;j++){
            // 가운데면 건너 뜀
            if(i == 1 && j == 1) continue;
            // 분할 후 정사각형 크기
            ll nxt = cur / 3;

            // 분할
            dnc(nxt, cy + i * nxt, cx + j * nxt);
        }
    }
}

int main(){
    cin >> n; dnc(n, 0, 0);
    for(int i = 0;i < n;i++){
        for(int j = 0;j < n;j++) cout << (a[i][j] ? '*' : ' ');
        cout << "\n";
    }

    return 0;
}